<a href="https://colab.research.google.com/github/RoBorregos/home-pipelines/blob/llm-train/Llama3_1_(8B)_GRPO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


# Set up drive directory for checkpoints

In [1]:
from google.colab import drive
import os

drive.mount('/content/drive')
checkpoint_dir = "/content/drive/MyDrive/RoBorregos/model_checkpoints/grpo"
os.makedirs(checkpoint_dir, exist_ok=True)
os.listdir(checkpoint_dir)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


[]

### News

**Read our [blog post](https://unsloth.ai/blog/r1-reasoning) for guidance on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [2]:
%%capture
# Skip restarting message in Colab
import sys; modules = list(sys.modules.keys())
for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None

!pip install unsloth==2025.3.18 unsloth_zoo==2025.3.16
!pip install vllm
!pip install pillow==11.1.0

### Unsloth

Use `PatchFastRL` before all functions to patch GRPO and other RL algorithms!

In [3]:
from unsloth import FastLanguageModel, PatchFastRL
PatchFastRL("GRPO", FastLanguageModel)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
Unsloth: Failed to patch Gemma3ForConditionalGeneration.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 04-15 19:40:19 [__init__.py:239] Automatically detected platform cuda.


Load up `Llama 3.1 8B Instruct`, and set parameters

In [6]:
from unsloth import is_bfloat16_supported
import torch
max_seq_length = 512 # Can increase for longer reasoning traces
lora_rank = 32 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "deepseek-ai/DeepSeek-R1-Distill-Llama-8B",
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.6, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ], # Remove QKVO if out of memory
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

==((====))==  Unsloth 2025.3.18: Fast Llama patching. Transformers: 4.51.1. vLLM: 0.8.4.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post2. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/deepseek-r1-distill-llama-8b-unsloth-bnb-4bit with actual GPU utilization = 59.43%
Unsloth: Your GPU has CUDA compute capability 7.5 with VRAM = 14.74 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 512. Num Sequences = 160.
Unsloth: vLLM's KV Cache can use up to 2.59 GB. Also swap space = 2 GB.
WARNING 04-15 19:43:20 [config.py:2836] Casting torch.bfloat16 to torch.float16.
INFO 04-15 19:43:35 [config.py:689] This model supports multiple tasks: {'classify', 'score', 'generat

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 04-15 19:45:46 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 04-15 19:45:46 [model_runner.py:1146] Model loading took 5.7737 GiB and 126.704938 seconds
INFO 04-15 19:45:58 [worker.py:267] Memory profiling takes 11.76 seconds
INFO 04-15 19:45:58 [worker.py:267] the current vLLM instance can use total_gpu_memory (14.74GiB) x gpu_memory_utilization (0.59) = 8.76GiB
INFO 04-15 19:45:58 [worker.py:267] model weights take 5.77GiB; non_torch_memory takes 0.03GiB; PyTorch activation peak memory takes 0.74GiB; the rest of the memory reserved for KV Cache is 2.22GiB.
INFO 04-15 19:45:59 [executor_base.py:112] # cuda blocks: 1136, # CPU blocks: 1024
INFO 04-15 19:45:59 [executor_base.py:117] Maximum concurrency for 512 tokens per request: 35.50x
INFO 04-15 19:46:00 [model_runner.py:1456] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI. If

Capturing CUDA graph shapes:   0%|          | 0/23 [00:00<?, ?it/s]

INFO 04-15 19:47:08 [model_runner.py:1598] Graph capturing finished in 67 secs, took 0.53 GiB
INFO 04-15 19:47:08 [llm_engine.py:449] init engine (profile, create kv cache, warmup model) took 81.37 seconds


Unsloth 2025.3.18 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


### Data Prep
<a name="Data"></a>

We directly leverage [@willccbb](https://gist.github.com/willccbb/4676755236bb08cab5f4e54a0475d6fb) for data prep and all reward functions. You are free to create your own!

In [23]:
import re
from datasets import load_dataset, Dataset
from typing import List, Optional
from pydantic import BaseModel, Field

# Load and prep dataset
SYSTEM_PROMPT = """
Respond in the following format:
<reasoning>
...
</reasoning>
<answer>
...
</answer>
"""

XML_COT_FORMAT = """\
<reasoning>
{reasoning}
</reasoning>
<answer>
{answer}
</answer>
"""

def extract_xml_answer(text: str) -> str:
    answer = text.split("<answer>")[-1]
    answer = answer.split("</answer>")[0]
    return answer.strip()

def extract_hash_answer(text: str) -> str | None:
    if "####" not in text:
        return None
    return text.split("####")[1].strip()

# uncomment middle messages for 1-shot prompting
def get_gsm8k_questions(split = "train") -> Dataset:
    data = load_dataset('openai/gsm8k', 'main')[split] # type: ignore
    data = data.map(lambda x: { # type: ignore
        'prompt': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': x['question']}
        ],
        'answer': extract_hash_answer(x['answer'])
    }) # type: ignore
    return data # type: ignore

import json

def load_custom_dataset(file_path):
    """
    Load custom dataset from JSON file and format for GRPO training
    """
    with open(file_path, 'r') as f:
        # Each line is a separate JSON object
        data = [json.loads(line) for line in f if line.strip()]

    # Create dataset dictionary with the format expected by GRPO trainer
    dataset_dict = {
        'prompt': [],
        'answer': []
    }

    for item in data:
        messages = item['messages']
        # The prompt is all messages except the last one (assistant's response)
        dataset_dict['prompt'].append(messages[:-1])
        # The answer is the content of the assistant's message
        dataset_dict['answer'].append(messages[-1]['content'])

    # Convert to HuggingFace dataset
    dataset = Dataset.from_dict(dataset_dict)
    return dataset

def load_gpsr_dataset(file_path):
    """
    Load custom dataset from JSON file and format for GRPO training
    """
    with open(file_path, 'r') as f:
        data = json.load(f)

    # Create dataset dictionary with the format expected by GRPO trainer
    dataset_dict = {
        'question': [],
        'prompt': [],
        'answer': []
    }

    for entry in data:
        dataset_dict['question'].append(entry['string_cmd'])
        dataset_dict['prompt'].append([
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': entry['string_cmd']}
        ])
        dataset_dict['answer'].append(entry['structured_cmd'])

    # Convert to HuggingFace dataset
    dataset = Dataset.from_dict(dataset_dict)
    return dataset

# dataset = get_gsm8k_questions()

# print(dataset)
# dataset = load_custom_dataset()
dataset = load_gpsr_dataset('dataset.json')

# Define expected response structure
class CommandShape(BaseModel):
    action: str = Field(description="The action to be performed")
    characteristic: Optional[str] = Field(description="A characteristic related to the action")
    complement: Optional[str] = Field(description="A complement related to the action")

class CommandListShape(BaseModel):
    commands: List[CommandShape]

available_actions = [
    "go","find_object","pick","find_person","find_person_by_name","count",
    "contextual_say","find_person_info","visual_info","ask_answer_question",
    "follow_person_until","guide_to","give","say","place"
]

def verify_response_shape(response: str) -> bool:
    """Check if response is a valid CommandListShape."""
    try:
        parsed = CommandListShape.model_validate_json(response)
        return isinstance(parsed, CommandListShape)
    except Exception:
        return False

def verify_soft_response_shape(response: str) -> bool:
    """Check if response contains something resembling commands."""
    return "commands" in response and "action" in response

def has_valid_actions(response: str) -> bool:
    """Check if the response contains only available actions."""
    try:
        parsed = CommandListShape.model_validate_json(response)
        for cmd in parsed.commands:
            if cmd.action not in available_actions:
                return False
        return True
    except Exception:
        return False

def has_characteristics_and_complements(response: str) -> bool:
    """Check if complement is present but characteristic is not"""
    try:
        parsed = CommandListShape.model_validate_json(response)
        for cmd in parsed.commands:
            if cmd.complement is not None and cmd.characteristic is None:
                return False
        return True
    except Exception:
        return False

# Combined reward functions
def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
    print(completions)
    print(completions[0])
    print(completions[0][0])
    contents = [completion[0]["content"] for completion in completions]
    extracted_contents = [extract_xml_answer(c) for c in contents]

    print(f"Completion: {extracted_contents[0]}, Answer: {answer}")

    return [2.0 if r == a else 0.0 for r, a in zip(extracted_contents, answer)]

def response_shape_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    extracted_contents = [extract_xml_answer(c) for c in contents]

    return [1.0 if verify_response_shape(c) else 0.0 for c in extracted_contents]

def soft_response_shape_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    extracted_contents = [extract_xml_answer(c) for c in contents]

    return [0.5 if verify_soft_response_shape(c) else 0.0 for c in extracted_contents]

def valid_action_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    extracted_contents = [extract_xml_answer(c) for c in contents]

    return [1.0 if has_valid_actions(c) else 0.0 for c in extracted_contents]

def characteristics_complements_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    extracted_contents = [extract_xml_answer(c) for c in contents]

    return [1.0 if has_characteristics_and_complements(c) else 0.0 for c in extracted_contents]

def strict_xml_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<reasoning>\n.*?\n</reasoning>\n<answer>\n.*?\n</answer>\n$"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def soft_xml_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def count_xml(text) -> float:
    count = 0.0
    if text.count("<reasoning>\n") == 1:
        count += 0.125
    if text.count("\n</reasoning>\n") == 1:
        count += 0.125
    if text.count("\n<answer>\n") == 1:
        count += 0.125
        count -= len(text.split("\n</answer>\n")[-1])*0.001
    if text.count("\n</answer>") == 1:
        count += 0.125
        count -= (len(text.split("\n</answer>")[-1]) - 1)*0.001
    return count

def xmlcount_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    return [count_xml(c) for c in contents]

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [8]:
from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    use_vllm = True, # use vLLM for fast inference!
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "paged_adamw_8bit",
    logging_steps = 1,
    bf16 = is_bfloat16_supported(),
    fp16 = not is_bfloat16_supported(),
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 6, # Decrease if out of memory
    max_prompt_length = 256,
    max_completion_length = 200,
    num_train_epochs = 1, # Set to 1 for a full training run
    # max_steps = 5,
    max_grad_norm = 0.1,
    report_to = "none", # Can use Weights & Biases
    output_dir = checkpoint_dir,
    save_strategy = "steps",
    save_steps = 50,
)

Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 6


[link text](https://)And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [ ]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        # response_shape_reward_func,
        # soft_response_shape_reward_func,
        # valid_action_reward_func,
        # characteristics_complements_reward_func,
        correctness_reward_func,
        response_shape_reward_func,
        soft_response_shape_reward_func,
        valid_action_reward_func,
        characteristics_complements_reward_func,
        strict_xml_format_reward_func,
        soft_xml_format_reward_func,
        xmlcount_reward_func,
    ],
    args = training_args,
    train_dataset = dataset,
)
try:
  trainer.train(resume_from_checkpoint = True)
except Exception as e:
  # Likely because of no checkpoints available
  print(e)
  trainer.train()

No valid checkpoint found in output directory (/content/drive/MyDrive/RoBorregos/model_checkpoints/grpo)


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 460 | Num Epochs = 1 | Total steps = 460
O^O/ \_/ \    Batch size per device = 6 | Gradient accumulation steps = 1
\        /    Data Parallel GPUs = 1 | Total batch size (6 x 1 x 1) = 6
 "-____-"     Trainable parameters = 83,886,080/8,000,000,000 (1.05% trained)


[[{'role': 'assistant', 'content': 'Alright, so I\'m trying to figure out how to answer the quiz from the person sitting to my left in the office. Hmm, first I should probably understand what the quiz is about. The message didn\'t specify, so I might need to make some assumptions. Since the person pointed me to respond, maybe they provided a specific question earlier that I need to address.\n\nWait, in the initial message, it just says "answer the quiz of the person pointing to the left in the office." I\'m a bit confused. Maybe they referred to a previous conversation or a shared document where the quiz was mentioned. Since I don\'t have access to that, perhaps I should prompt them to provide more details.\n\nI should consider if the quiz is about general knowledge, perhaps a puzzle, or something related to our work. Maybe it\'s about office procedures, team trivia, or something fun. I don\'t want to assume too much and give the wrong answer. It\'s better to ask for clarification to e

[[{'role': 'assistant', 'content': "Alright, so I need to figure out how to help the user with their request. They want me to find a snack in the office and bring it to the person sitting in the living room. Hmm, okay, let me break this down.\n\nFirst, I should think about the steps involved. Find a snack in the office. That means I need to know where the snacks are located. Usually, snacks are kept in a kitchen or a pantry. I should check those areas. Maybe the user can tell me where the snacks are stored. Oh, wait, the user didn't specify that they have a specific type of snack in mind. So I might need to ask them about their preference to choose the right snack.\n\nNext, once I find the snack, I need to fetch it. That's straightforward; I just need to go over to the office and pick it up. But wait, am I allowed to go into the kitchen or office? I should confirm whether the area where the snacks are"}], [{'role': 'assistant', 'content': 'Okay, let me try to figure out how to solve th

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [ ]:
text = tokenizer.apply_chat_template([
    {"role" : "user", "content" : "Calculate pi."},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

In [ ]:
dataset['question']


And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [ ]:
model.save_lora("grpo_saved_lora")

Now we load the LoRA and test:

In [ ]:
text = tokenizer.apply_chat_template([
    {"role" : "system", "content" : SYSTEM_PROMPT},
    {"role" : "user", "content" : "Calculate pi."},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/drive/1WZDi7APtQ9VsvOrQSSC5DDtxq159j8iZ?usp=sharing)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if True: model.push_to_hub_gguf("diegohc/finetuning-tests", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Llama 3.2 Conversational notebook. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(1B_and_3B)-Conversational.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
